### PR TITLE
Added check for maximum size of downloaded file names

### LINKED ISSUE #1808: Package does not install due to the file name exceeding the maximum amount of characters
### Brief description of your issue

image

Logs from the `DiagOutputDir` folder with `--verbose-logs`:
1) `winget validate` = https://pastebin.com/raw/NxWVKnvg
2) `winget install` = https://pastebin.com/raw/QdchskSe

### Steps to reproduce

Install the WinGet package from https://github.com/microsoft/winget-pkgs/pull/37971 

^ https://github.com/Dragon1573/winget-pkgs/tree/bf22a2a5a907bef7cf7c4d00595ede8745b7e84d/manifests/h/Huawei/HiSuite/11.0.0.580

### Expected behavior

For the WinGet package to install successfully on my PC.

### Actual behavior

Due to how weird the file name is for the installer, it has seemingly exceeded the maximum amount of characters you can write in a file name.

Perhaps if WinGet spots a file name that is too long, it should possibly truncate it to avoid this issue?

### Environment

```shell
Windows Package Manager (Preview) v1.2.3411-preview
Copyright (c) Microsoft Corporation. All rights reserved.

Windows: Windows.Desktop v10.0.22523.1000
Package: Microsoft.DesktopAppInstaller v1.17.3411.0

Logs: %LOCALAPPDATA%\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\DiagOutputDir

Links
---------------------------------------------------------------------------
Privacy Statement   https://aka.ms/winget-privacy
License Agreement   https://aka.ms/winget-license
Third Party Notices https://aka.ms/winget-3rdPartyNotice
Homepage            https://aka.ms/winget
Windows Store Terms https://www.microsoft.com/en-us/storedocs/terms-of-sale
```

### PR DESCRIPTION
**Pull request by jedieaston**
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [x] Are you working against an Issue?

-----
Resolves #1808.

This PR adds a check to GetInstallerPostHashValidationFileName that ensures that the filename in the URL does not exceed 260 characters (the maximum ~~path size~~ file name size in Windows). If it does, then the "older" naming style (`manifest.version.ext`) is used for the downloaded file. Technically this could still fail if the full length of the path is 260 instead of just the file name, but I thought that was fairly unlikely. If it isn't I can adjust for that too.

Tested: manually via the manifest given in the issue. As always, I'm happy to add tests if needed.

Edit: corrected explanation, thanks.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/1842)

### CODE REVIEW

File: src/AppInstallerCLICore/Workflows/DownloadFlow.cpp [hunk 1]
```diff
@@ -64,9 +64,9 @@ namespace AppInstaller::CLI::Workflow
             // Get file name from download URI
             std::filesystem::path filename = GetFileNameFromURI(context.Get<Execution::Data::Installer>()->Url);
 
-            // Assuming that we find a stem value in the URI, use it.
+            // Assuming that we find a safe stem value in the URI, use it.
             // This should be extremely common, but just in case fall back to the older name style.
-            if (filename.has_stem())
+            if (filename.has_stem() && filename.string().size() < MAX_PATH)
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Workflows/DownloadFlow.cpp:69]:
```suggestion
            if (filename.has_stem() && filename.string().size() + GetInstallerFileExtension(context).size() < MAX_PATH)
```
Probably would actually want to get and hold the extension in a local before this so we don't do it twice.

**Comment by JohnMcPMS** [comment]
Is this something you are seeing in practice?

File: src/AppInstallerCLICore/Workflows/DownloadFlow.cpp (hunk 1 above)
Comment (jedieaston) [reply src/AppInstallerCLICore/Workflows/DownloadFlow.cpp]: Yep, that makes sense. Fixed!

**Review by JohnMcPMS — APPROVED** [review]

**Pull request merged**